### PR TITLE
Send email notifications when invitations are accepted/confirmed

### PR DESCRIPTION
**Pull request by njfox**
This PR sends email notifications to users when their invitee accepts an invitation so they know when to confirm them. It also sends an email to the invitee when they are confirmed, which is the behavior indicated by the web vault when they call the /accept endpoint.

Additionally, I've replaced the specific `send_invite` function with a generic email sending function. This requires the subject and body to be formatted by the caller instead of the callee, which I think offers some flexibility for sending many different types of email, but I am open to discussion on whether we want to keep this approach.

### CODE REVIEW

File: src/api/admin.rs [hunk 1]
```diff
@@ -43,7 +45,21 @@ fn invite_user(data: JsonUpcase<InviteData>, _token: AdminToken, conn: DbConn) -
...
+        let claims = generate_invite_claims(
+            user.uuid.to_string(),
+            user.email.clone(),
+            org_id.clone(),
+            None,
+            None,
+        );
+        let org_name = "bitwarden_rs";
+        let invite_token = encode_jwt(&claims);
+        mail::send_invite(&user.email, &org_id, &user.uuid, &invite_token, &org_name, mail_config)?;
```
Comment (dani-garcia) [inline src/api/admin.rs]: The generate_claims method receives pretty much the same parameters as the send_invite, and it's value is only used there, maybe we should just move the generate_claims call inside the send_invite?

Also, I don't really like the idea of using magic values for the org_id like that, maybe we can change it to optional like the user_org_id after it? We'll need to provide a default value for the vault URL with `unwrap_or`, if you set it to `_`, does the web vault complain? The link is already big enough, and that fake id is both a url parameter and part of the token, so we could save some space.

File: src/api/core/organizations.rs [hunk 2]
```diff
@@ -617,6 +604,20 @@ fn accept_invite(_org_id: String, _org_user_id: String, data: JsonUpcase<AcceptD
         None => err!("Invited user not found"),
     }
 
+    if let Some(ref mail_config) = CONFIG.mail {
+        let org_name = match Organization::find_by_uuid(&claims.org_id, &conn) {
+                Some(org) => org.name,
+                None => String::from("bitwarden_rs"),
+        };
+        if claims.invited_by_email.is_some() {
```
Comment (dani-garcia) [inline src/api/core/organizations.rs]: This is more clear as `if let Some(invited_by_email) = claims.invited_by_email { ...`

File: src/mail.rs [hunk 3]
```diff
@@ -87,17 +76,54 @@ pub fn send_invite(
...
+    new_user_email: &str,
+    address: &str,
+    org_name: &str,
+    config: &MailConfig,
+) -> EmptyResult {
+    let (subject, body) = {
+        ("Invitation accepted",
+        format!(
+            "<html>
+             <p>Your invitation to <b>{}</b> to join <b>{}</b> was accepted. Please log in to the bitwarden_rs server and confirm them from the organization management page.</p>
```
Comment (dani-garcia) [inline src/mail.rs]: Wouldn't it make more sense for it to say:
Your invitation **for** `email` to join `org` was accepted

File: src/mail.rs [hunk 4]
```diff
@@ -87,17 +76,54 @@ pub fn send_invite(
...
+    let email = EmailBuilder::new()
+    .to(address)
+    .from((config.smtp_from.clone(), "Bitwarden-rs"))
+    .subject(subject)
+    .header(("Content-Type", "text/html"))
+    .body(body)
+    .build()
+    .map_err(|e| Error::new("Error building email", e.to_string()))?;
+
+mailer(config)
```
Comment (dani-garcia) [inline src/mail.rs]: This is formatted strangely, it looks like it's out of the function

Comment (njfox) [reply src/mail.rs]: Oops, good catch. Fixed in 2f5bdc2.

File: src/mail.rs (hunk 3 above)
Comment (njfox) [reply src/mail.rs]: Fixed in 2f5bdc2.

File: src/mail.rs [hunk 5]
```diff
@@ -87,17 +76,54 @@ pub fn send_invite(
...
+pub fn send_invite_confirmed(
+    address: &str,
+    org_name: &str,
+    config: &MailConfig,
+) -> EmptyResult {
+    let (subject, body) = {
+        (format!("Invitation to {} confirmed", org_name),
+        format!(
+            "<html>
+             <p>Your invitation to join <b>{}</b> was accepted. It will now appear under the Organizations the next time you log into the web vault.</p>
```
[the remark this answers is not in the document]
Comment (njfox) [reply src/mail.rs]: Agreed, fixed in 2f5bdc2.

File: src/api/core/organizations.rs (hunk 2 above)
Comment (njfox) [reply src/api/core/organizations.rs]: Yeah, I was originally going to do that but I often get a little confused about the `if let Some()` syntax. Thanks for the example--I added this in 7db66f7.

File: src/api/admin.rs (hunk 1 above)
Comment (njfox) [reply src/api/admin.rs]: > Also, I don't really like the idea of using magic values for the org_id like that, maybe we can change it to optional like the user_org_id after it? We'll need to provide a default value for the vault URL with unwrap_or, if you set it to _, does the web vault complain?

I agree, I don't like using magic values either. Just to make sure I understand, are you suggesting we replace the `org_id` URL parameter with `_` in the invite links?

Comment (dani-garcia) [reply src/api/admin.rs]: Yeah, in the struct we can have an optional value, but in the URL, we'll need something for the case when org_id is none. With `_`, the URL would end up like `/#/accept-organization/?organizationId=_&organizationUserId...`.

We could try removing the query parameter from the URL completely, but that seems like too much work for something that may not even work. Using an empty string may work though, if the web vault doesn't choke on it.

Comment (njfox) [reply src/api/admin.rs]: Good idea. I've implemented this in 0a74e79 and gone through the normal org invite flow as well as inviting from the /admin page and both seem to be working fine.

**Comment by dani-garcia** [comment]
Great, thanks @njfox!

**Pull request merged**